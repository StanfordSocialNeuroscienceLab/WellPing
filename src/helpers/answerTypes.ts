import { AnswerEntity } from "../entities/AnswerEntity";
import { Question, QuestionsList, Choice } from "./types";

// We use `{ value: ... }` because it seems that `"simple-json"`
// does not support simply mixing normal values.
// This also increases the consistency of the data.

export type SliderAnswerData = {
  value: number;
};

export type ChoicesWithSingleAnswerAnswerData = {
  value: Choice;
};

// We use an array of tuples here (instead of `{ [key: string]: boolean }`)
// because the order shown on the screen can be kept in the results.
// This is potentially helpful for analysis later if the order is randomized
// as now we know the order the user saw the choices in.
export type ChoicesWithMultipleAnswersAnswerChoices = [Choice, boolean][];
export type ChoicesWithMultipleAnswersAnswerData = {
  value: ChoicesWithMultipleAnswersAnswerChoices;
};

export type YesNoAnswerData = {
  value: boolean;
};

export type MultipleTextAnswerData = {
  value: string[];
};

export type HowLongAgoAnswerDataType = [number | null, string | null];
export type HowLongAgoAnswerData = {
  value: HowLongAgoAnswerDataType;
};

export type AnswerData =
  | SliderAnswerData
  | ChoicesWithSingleAnswerAnswerData
  | ChoicesWithMultipleAnswersAnswerData
  | YesNoAnswerData
  | MultipleTextAnswerData
  | HowLongAgoAnswerData;

export interface AnswersList {
  [id: string]: AnswerEntity;
}

export interface QuestionScreenProps {
  question: Question;
  onDataChange: (data: AnswerData) => void;
  allAnswers: AnswersList;
  allQuestions: QuestionsList;
  pipeInExtraMetaData: (input: string) => string;
  setDataValidationFunction: (func: () => boolean) => void;
}
