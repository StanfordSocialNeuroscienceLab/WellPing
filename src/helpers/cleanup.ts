import { secureRemoveAllAnswersAsync } from "./secureStore/answer";
import { secureRemoveAllPingsAsync } from "./secureStore/ping";

export async function clearAllPingsAndAnswersAsync(): Promise<void> {
  await secureRemoveAllAnswersAsync();
  await secureRemoveAllPingsAsync();
}
