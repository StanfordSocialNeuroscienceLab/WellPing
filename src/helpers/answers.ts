import { AnswerData, Answer } from "./answerTypes";
import { getAnswersPingIdsQuestionIdsListAsync } from "./asyncStorage/answersPingIdsQuestionIdsList";
import { AnswerSchema } from "./schemas/Answer";
import {
  secureGetAnswerAsync,
  secureStoreAnswerAsync,
} from "./secureStore/answer";
import { Question, Ping } from "./types";

export async function getAnswersAsync(): Promise<Answer[]> {
  const answersList = await getAnswersPingIdsQuestionIdsListAsync();

  // https://stackoverflow.com/q/28066429/2603230
  const answers: Answer[] = await Promise.all(
    answersList.map(async (pingIdAndQuestionId) => {
      const answer = await secureGetAnswerAsync(
        pingIdAndQuestionId[0],
        pingIdAndQuestionId[1],
      );
      if (answer === null) {
        throw new Error(
          "answer === null in answersList.map in getAnswersAsync.",
        );
      }
      return answer;
    }),
  );

  return answers;
}

export async function insertAnswerAsync({
  ping,
  question,
  realQuestionId,
  preferNotToAnswer,
  data,
  date,
}: {
  ping: Ping;
  question: Question;
  realQuestionId: string;
  preferNotToAnswer: true | null; // See `MARK: WHY_PNA_TRUE_OR_NULL`.
  data: AnswerData | null;
  date: Date;
}): Promise<Answer> {
  const answer = AnswerSchema.parse({
    pingId: ping.id,
    questionId: realQuestionId,
    preferNotToAnswer,
    data,
    date,
  });

  await secureStoreAnswerAsync(answer);

  return answer;
}
