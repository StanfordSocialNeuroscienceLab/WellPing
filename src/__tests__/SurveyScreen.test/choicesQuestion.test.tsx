/**
 * Notice that `randomizeChoicesOrder` and `randomizeExceptForChoiceIds` are not
 * tested here as they are tested in ChoicesQuestionScreen.test.tsx.
 */
import React from "react";
import {
  render,
  fireEvent,
  waitFor,
  RenderAPI,
} from "react-native-testing-library";

import SurveyScreen, { SurveyScreenProps } from "../../SurveyScreen";
import { QuestionType } from "../../helpers/helpers";
import {
  TEST_PING,
  mockDatabaseRelatedFunction,
  testQuestionsSequenceAsync,
} from "./helper";

beforeEach(() => {
  mockDatabaseRelatedFunction();
});

async function clickOptionAsync(
  option: string,
  { findByA11yLabel }: RenderAPI,
) {
  // `findBy` does the `waitFor` for us.
  fireEvent.press(await findByA11yLabel(`select ${option}`));
}

const propsBase: SurveyScreenProps = {
  questions: {},
  startingQuestionId: "q1",
  ping: TEST_PING,
  previousState: null,
  onFinish: async () => {},
};

const TWO_TYPES: (
  | "ChoicesWithSingleAnswer"
  | "ChoicesWithMultipleAnswers"
)[] = [
  QuestionType.ChoicesWithSingleAnswer,
  QuestionType.ChoicesWithMultipleAnswers,
];

describe.each(TWO_TYPES)("simple test - %s: ", (type) => {
  // All these should not change the next question.
  describe.each([
    undefined,
    [],
    [["Non Existent Choice", "newId"]] as [string, string][],
  ])("specialCasesStartId: %p", (specialCasesStartId) => {
    const props: SurveyScreenProps = {
      ...propsBase,
      questions: {
        q1: {
          id: "q1",
          type,
          specialCasesStartId,
          question: "Question 1",
          choices: ["Choice 1", "Choice 2", "Choice 3"],
          next: "q2",
        },
        q2: {
          id: "q2",
          type: QuestionType.Slider,
          question: "Question 2",
          slider: ["left", "right"],
          next: null,
        },
      },
    };

    test("click a choice", async () => {
      const onFinishFn = jest.fn();

      const renderResults = render(
        <SurveyScreen {...props} onFinish={onFinishFn} />,
      );

      await testQuestionsSequenceAsync({
        renderResults,
        onFinishFn,
        sequence: [
          {
            expectCurrentQuestionAsync: async (getCurrentQuestionTitle) => {
              await clickOptionAsync("Choice 2", renderResults);

              expect(getCurrentQuestionTitle()).toBe("Question 1");
            },
          },
          {
            expectCurrentQuestionAsync: async (getCurrentQuestionTitle) => {
              expect(getCurrentQuestionTitle()).toBe("Question 2");
            },
          },
        ],
      });
    });

    test("click next without answering", async () => {
      const onFinishFn = jest.fn();

      const renderResults = render(
        <SurveyScreen {...props} onFinish={onFinishFn} />,
      );

      await testQuestionsSequenceAsync({
        renderResults,
        onFinishFn,
        sequence: [
          {
            expectCurrentQuestionAsync: async (getCurrentQuestionTitle) => {
              expect(getCurrentQuestionTitle()).toBe("Question 1");
            },
          },
          {
            expectCurrentQuestionAsync: async (getCurrentQuestionTitle) => {
              expect(getCurrentQuestionTitle()).toBe("Question 2");
            },
          },
        ],
      });
    });

    test("click prefer not to answer", async () => {
      const onFinishFn = jest.fn();

      const renderResults = render(
        <SurveyScreen {...props} onFinish={onFinishFn} />,
      );

      await testQuestionsSequenceAsync({
        renderResults,
        onFinishFn,
        sequence: [
          {
            expectCurrentQuestionAsync: async (getCurrentQuestionTitle) => {
              // Clicking a choice shouldn't matter here if we click prefer not
              // to answer.
              await clickOptionAsync("Choice 1", renderResults);

              expect(getCurrentQuestionTitle()).toBe("Question 1");
            },
            nextButton: "pna",
          },
          {
            expectCurrentQuestionAsync: async (getCurrentQuestionTitle) => {
              expect(getCurrentQuestionTitle()).toBe("Question 2");
            },
          },
        ],
      });
    });
  });
});
