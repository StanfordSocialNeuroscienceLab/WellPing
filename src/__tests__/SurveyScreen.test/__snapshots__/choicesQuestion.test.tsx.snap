// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`with non-null specialCasesStartId - ChoicesWithMultipleAnswers:  click next without answering: getAnswersAsync 1`] = `
Array [
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q1",
  },
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q2",
  },
]
`;

exports[`with non-null specialCasesStartId - ChoicesWithMultipleAnswers:  click no-special-case choice: getAnswersAsync 1`] = `
Array [
  Object {
    "data": Object {
      "value": Array [
        Array [
          "Choice 1",
          false,
        ],
        Array [
          "Choice 2",
          true,
        ],
        Array [
          "Choice 3",
          false,
        ],
      ],
    },
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q1",
  },
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q2",
  },
]
`;

exports[`with non-null specialCasesStartId - ChoicesWithMultipleAnswers:  click prefer not to answer: getAnswersAsync 1`] = `
Array [
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": true,
    "questionId": "q1",
  },
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q2",
  },
]
`;

exports[`with non-null specialCasesStartId - ChoicesWithMultipleAnswers:  click special-case choice: getAnswersAsync 1`] = `
Array [
  Object {
    "data": Object {
      "value": Array [
        Array [
          "Choice 1",
          true,
        ],
        Array [
          "Choice 2",
          false,
        ],
        Array [
          "Choice 3",
          false,
        ],
      ],
    },
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q1",
  },
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q1_c1",
  },
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q2",
  },
]
`;

exports[`with non-null specialCasesStartId - ChoicesWithMultipleAnswers:  click two special-case choices: getAnswersAsync 1`] = `
Array [
  Object {
    "data": Object {
      "value": Array [
        Array [
          "Choice 1",
          true,
        ],
        Array [
          "Choice 2",
          false,
        ],
        Array [
          "Choice 3",
          true,
        ],
      ],
    },
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q1",
  },
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q1_c1",
  },
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q2",
  },
]
`;

exports[`with non-null specialCasesStartId - ChoicesWithSingleAnswer:  click next without answering: getAnswersAsync 1`] = `
Array [
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q1",
  },
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q2",
  },
]
`;

exports[`with non-null specialCasesStartId - ChoicesWithSingleAnswer:  click no-special-case choice: getAnswersAsync 1`] = `
Array [
  Object {
    "data": Object {
      "value": "Choice 2",
    },
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q1",
  },
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q2",
  },
]
`;

exports[`with non-null specialCasesStartId - ChoicesWithSingleAnswer:  click prefer not to answer: getAnswersAsync 1`] = `
Array [
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": true,
    "questionId": "q1",
  },
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q2",
  },
]
`;

exports[`with non-null specialCasesStartId - ChoicesWithSingleAnswer:  click special-case choice: getAnswersAsync 1`] = `
Array [
  Object {
    "data": Object {
      "value": "Choice 1",
    },
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q1",
  },
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q1_c1",
  },
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q2",
  },
]
`;

exports[`with non-null specialCasesStartId - ChoicesWithSingleAnswer:  click two special-case choices: getAnswersAsync 1`] = `
Array [
  Object {
    "data": Object {
      "value": "Choice 3",
    },
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q1",
  },
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q1_c3",
  },
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q2",
  },
]
`;

exports[`with some null specialCasesStartId - ChoicesWithMultipleAnswers:  click next without answering: getAnswersAsync 1`] = `
Array [
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q1",
  },
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q2",
  },
]
`;

exports[`with some null specialCasesStartId - ChoicesWithMultipleAnswers:  click no-special-case choice: getAnswersAsync 1`] = `
Array [
  Object {
    "data": Object {
      "value": Array [
        Array [
          "Choice 1",
          false,
        ],
        Array [
          "Choice 2",
          true,
        ],
        Array [
          "Choice 3",
          false,
        ],
      ],
    },
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q1",
  },
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q2",
  },
]
`;

exports[`with some null specialCasesStartId - ChoicesWithMultipleAnswers:  click prefer not to answer: getAnswersAsync 1`] = `
Array [
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": true,
    "questionId": "q1",
  },
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q2",
  },
]
`;

exports[`with some null specialCasesStartId - ChoicesWithMultipleAnswers:  click special-case choice with null: getAnswersAsync 1`] = `
Array [
  Object {
    "data": Object {
      "value": Array [
        Array [
          "Choice 1",
          false,
        ],
        Array [
          "Choice 2",
          false,
        ],
        Array [
          "Choice 3",
          true,
        ],
      ],
    },
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q1",
  },
]
`;

exports[`with some null specialCasesStartId - ChoicesWithMultipleAnswers:  click special-case choice without null: getAnswersAsync 1`] = `
Array [
  Object {
    "data": Object {
      "value": Array [
        Array [
          "Choice 1",
          true,
        ],
        Array [
          "Choice 2",
          false,
        ],
        Array [
          "Choice 3",
          false,
        ],
      ],
    },
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q1",
  },
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q1_c1",
  },
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q2",
  },
]
`;

exports[`with some null specialCasesStartId - ChoicesWithMultipleAnswers:  click two special-case choices (with the latter one being null): getAnswersAsync 1`] = `
Array [
  Object {
    "data": Object {
      "value": Array [
        Array [
          "Choice 1",
          true,
        ],
        Array [
          "Choice 2",
          false,
        ],
        Array [
          "Choice 3",
          true,
        ],
      ],
    },
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q1",
  },
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q1_c1",
  },
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q2",
  },
]
`;

exports[`with some null specialCasesStartId - ChoicesWithSingleAnswer:  click next without answering: getAnswersAsync 1`] = `
Array [
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q1",
  },
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q2",
  },
]
`;

exports[`with some null specialCasesStartId - ChoicesWithSingleAnswer:  click no-special-case choice: getAnswersAsync 1`] = `
Array [
  Object {
    "data": Object {
      "value": "Choice 2",
    },
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q1",
  },
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q2",
  },
]
`;

exports[`with some null specialCasesStartId - ChoicesWithSingleAnswer:  click prefer not to answer: getAnswersAsync 1`] = `
Array [
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": true,
    "questionId": "q1",
  },
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q2",
  },
]
`;

exports[`with some null specialCasesStartId - ChoicesWithSingleAnswer:  click special-case choice with null: getAnswersAsync 1`] = `
Array [
  Object {
    "data": Object {
      "value": "Choice 3",
    },
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q1",
  },
]
`;

exports[`with some null specialCasesStartId - ChoicesWithSingleAnswer:  click special-case choice without null: getAnswersAsync 1`] = `
Array [
  Object {
    "data": Object {
      "value": "Choice 1",
    },
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q1",
  },
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q1_c1",
  },
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q2",
  },
]
`;

exports[`with some null specialCasesStartId - ChoicesWithSingleAnswer:  click two special-case choices (with the latter one being null): getAnswersAsync 1`] = `
Array [
  Object {
    "data": Object {
      "value": "Choice 3",
    },
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q1",
  },
]
`;

exports[`without meaningful specialCasesStartId - ChoicesWithMultipleAnswers:  specialCasesStartId: [[Array]] click a choice: getAnswersAsync 1`] = `
Array [
  Object {
    "data": Object {
      "value": Array [
        Array [
          "Choice 1",
          false,
        ],
        Array [
          "Choice 2",
          true,
        ],
        Array [
          "Choice 3",
          false,
        ],
      ],
    },
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q1",
  },
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q2",
  },
]
`;

exports[`without meaningful specialCasesStartId - ChoicesWithMultipleAnswers:  specialCasesStartId: [[Array]] click next without answering: getAnswersAsync 1`] = `
Array [
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q1",
  },
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q2",
  },
]
`;

exports[`without meaningful specialCasesStartId - ChoicesWithMultipleAnswers:  specialCasesStartId: [[Array]] click prefer not to answer: getAnswersAsync 1`] = `
Array [
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": true,
    "questionId": "q1",
  },
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q2",
  },
]
`;

exports[`without meaningful specialCasesStartId - ChoicesWithMultipleAnswers:  specialCasesStartId: [] click a choice: getAnswersAsync 1`] = `
Array [
  Object {
    "data": Object {
      "value": Array [
        Array [
          "Choice 1",
          false,
        ],
        Array [
          "Choice 2",
          true,
        ],
        Array [
          "Choice 3",
          false,
        ],
      ],
    },
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q1",
  },
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q2",
  },
]
`;

exports[`without meaningful specialCasesStartId - ChoicesWithMultipleAnswers:  specialCasesStartId: [] click next without answering: getAnswersAsync 1`] = `
Array [
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q1",
  },
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q2",
  },
]
`;

exports[`without meaningful specialCasesStartId - ChoicesWithMultipleAnswers:  specialCasesStartId: [] click prefer not to answer: getAnswersAsync 1`] = `
Array [
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": true,
    "questionId": "q1",
  },
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q2",
  },
]
`;

exports[`without meaningful specialCasesStartId - ChoicesWithMultipleAnswers:  specialCasesStartId: undefined click a choice: getAnswersAsync 1`] = `
Array [
  Object {
    "data": Object {
      "value": Array [
        Array [
          "Choice 1",
          false,
        ],
        Array [
          "Choice 2",
          true,
        ],
        Array [
          "Choice 3",
          false,
        ],
      ],
    },
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q1",
  },
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q2",
  },
]
`;

exports[`without meaningful specialCasesStartId - ChoicesWithMultipleAnswers:  specialCasesStartId: undefined click next without answering: getAnswersAsync 1`] = `
Array [
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q1",
  },
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q2",
  },
]
`;

exports[`without meaningful specialCasesStartId - ChoicesWithMultipleAnswers:  specialCasesStartId: undefined click prefer not to answer: getAnswersAsync 1`] = `
Array [
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": true,
    "questionId": "q1",
  },
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q2",
  },
]
`;

exports[`without meaningful specialCasesStartId - ChoicesWithSingleAnswer:  specialCasesStartId: [[Array]] click a choice: getAnswersAsync 1`] = `
Array [
  Object {
    "data": Object {
      "value": "Choice 2",
    },
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q1",
  },
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q2",
  },
]
`;

exports[`without meaningful specialCasesStartId - ChoicesWithSingleAnswer:  specialCasesStartId: [[Array]] click next without answering: getAnswersAsync 1`] = `
Array [
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q1",
  },
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q2",
  },
]
`;

exports[`without meaningful specialCasesStartId - ChoicesWithSingleAnswer:  specialCasesStartId: [[Array]] click prefer not to answer: getAnswersAsync 1`] = `
Array [
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": true,
    "questionId": "q1",
  },
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q2",
  },
]
`;

exports[`without meaningful specialCasesStartId - ChoicesWithSingleAnswer:  specialCasesStartId: [] click a choice: getAnswersAsync 1`] = `
Array [
  Object {
    "data": Object {
      "value": "Choice 2",
    },
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q1",
  },
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q2",
  },
]
`;

exports[`without meaningful specialCasesStartId - ChoicesWithSingleAnswer:  specialCasesStartId: [] click next without answering: getAnswersAsync 1`] = `
Array [
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q1",
  },
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q2",
  },
]
`;

exports[`without meaningful specialCasesStartId - ChoicesWithSingleAnswer:  specialCasesStartId: [] click prefer not to answer: getAnswersAsync 1`] = `
Array [
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": true,
    "questionId": "q1",
  },
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q2",
  },
]
`;

exports[`without meaningful specialCasesStartId - ChoicesWithSingleAnswer:  specialCasesStartId: undefined click a choice: getAnswersAsync 1`] = `
Array [
  Object {
    "data": Object {
      "value": "Choice 2",
    },
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q1",
  },
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q2",
  },
]
`;

exports[`without meaningful specialCasesStartId - ChoicesWithSingleAnswer:  specialCasesStartId: undefined click next without answering: getAnswersAsync 1`] = `
Array [
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q1",
  },
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q2",
  },
]
`;

exports[`without meaningful specialCasesStartId - ChoicesWithSingleAnswer:  specialCasesStartId: undefined click prefer not to answer: getAnswersAsync 1`] = `
Array [
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": true,
    "questionId": "q1",
  },
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q2",
  },
]
`;
